execution_engine/src/storage/trie_store/operations/mod.rs

```
@@ -505,41 +512,167 @@ where
+pub fn delete<'a, K, V, R, S, E>(
     correlation_id: CorrelationId,
-    txn: &mut T,
+    environment: &'a R,
     store: &S,
     root: &Digest,
     key_to_delete: &K,
 ) -> Result<DeleteResult, E>
 where
+    R: TransactionSource<'a, Handle = S::Handle>,
     K: ToBytes + FromBytes + Clone + PartialEq + std::fmt::Debug,
     V: ToBytes + FromBytes + Clone,
-    T: Readable<Handle = S::Handle> + Writable<Handle = S::Handle>,
     S: TrieStore<K, V>,
-    S::Error: From<T::Error>,
-    E: From<S::Error> + From<bytesrepr::Error>,
+    S::Error: From<R::Error>,
+    E: From<R::Error> + From<S::Error> + From<bytesrepr::Error>,
 {
-    let root_trie = match store.get(txn, root)? {
+    let mut txn = environment.create_read_write_txn()?;
+    let root_trie_bytes = match store.get_raw(&txn, root)? {
         None => return Ok(DeleteResult::RootNotFound),
         Some(root_trie) => root_trie,
     };
 
     let key_bytes = key_to_delete.to_bytes()?;
-    let TrieScan { tip, mut parents } =
-        scan::<_, _, _, _, E>(correlation_id, txn, store, &key_bytes, &root_trie)?;
+    let TrieScanRaw { tip, mut parents } =
+        scan_raw::<_, _, _, _, E>(correlation_id, &txn, store, &key_bytes, root_trie_bytes)?;
 
     // Check that tip is a leaf
     match tip {
-        Trie::Leaf { key, .. } if key == *key_to_delete => {}
+        Either::Left(bytes)
+            if {
+                // Partially deserialize a key of a leaf node to ensure that we can only continue if
+                // the key matches what we're looking for.
+                let ((tag_u8, key), _rem): ((u8, K), _) = FromBytes::from_bytes(&bytes)?;
+                // _rem contains bytes of serialized V, but we don't need to inspect it.
+                assert_eq!(
+                    TrieTag::from_u8(tag_u8),
+                    Some(TrieTag::Leaf),
+                    "Tip contains Leaf bytes"
+                );
```

We should return an error, as this will kill a node.

---

`Either::Left` branch here implies that this tag is a `TrieTag::Leaf`, if for some reason non-leafs are inside `Either::Left` then it's a programmer's error. Here's the relevant code from `scan_raw` where we also inspect first byte: https://github.com/casper-network/casper-node/blob/6233f31fcb3eebd188b6e020f39786cd89fb2f25/execution_engine/src/storage/trie_store/operations/mod.rs#L551-L556